PySDM/dynamics/seeding.py

```diff
@@ -0,0 +1,38 @@
+""" particle injection handling, requires initalising a simulation with
+enough particles flagged with NaN multiplicity (translated to zeros
+at multiplicity discretisation """
+
+
+class Seeding:
+    def __init__(
+        self,
+        *,
+        time_window: tuple,
+        seeded_particle_extensive_attributes: dict,
+        seeded_particle_multiplicity: int,
+    ):
+        self.particulator = None
+        self.time_window = time_window
+        self.seeded_particle_extensive_attributes = seeded_particle_extensive_attributes
+        self.seeded_particle_multiplicity = seeded_particle_multiplicity
+
+    def register(self, builder):
+        self.particulator = builder.particulator
+
+    def __call__(self):
+        if self.particulator.n_steps == 0:
+            if tuple(
+                self.particulator.attributes.get_extensive_attribute_keys()
+            ) != tuple(self.seeded_particle_extensive_attributes.keys()):
+                raise ValueError(
+                    f"extensive attributes ({self.seeded_particle_extensive_attributes.keys()})"
+                    " do not match those used in particulator"
+                    f" ({self.particulator.attributes.get_extensive_attribute_keys()})"
+                )
+
+        time = self.particulator.n_steps * self.particulator.dt
+        if self.time_window[0] <= time <= self.time_window[1]:
```

Comment:
Re: my comment in https://github.com/open-atmos/PySDM/issues/1331 again -- Maybe a possible way to update this logic to use a variable injection rate, rather than a fixed injection rate in a given time window, would be to change this if statement to something like `if self.injection_rate > 0` and then scale the `self.seeded_particle_multiplicity` by the injection rate?

Reply:
I can experiment with this modification!

Reply:
thanks @jtbuch & @claresinger, here's a starting point which might help: https://github.com/open-atmos/PySDM/pull/1367/commits/6633c2b7766994baeaa76dce5c5d9d046ccf28a6